File: framework/lib/rect_shape.dart
part of uxml;

/**
 * Define a rectangular shape.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class RectShape extends Shape {
  static ElementDef rectshapeElementDef;
  /** Border radius property definition */
  static PropertyDefinition borderRadiusProperty;

  RectShape() :super() {
  }

  /**
   * Sets or returns border radius.
   */
  BorderRadius get borderRadius {
    return getProperty(borderRadiusProperty);
  }

  void set borderRadius(BorderRadius value) {
    setProperty(borderRadiusProperty, value);
  }

  /** Overrides UIElement.onRedraw. */
  void onRedraw(UISurface surface) {
    // TODO(ferhat): abstract away the use of setBackground vs drawRect
    // inside ui surfaces instead of redraw.
    bool requiresCanvas = false;
    if (hasFilters) {
      Filters f = filters;
      if (f._innerShadow != null && f._innerShadow.knockout) {
        requiresCanvas = true;
      } else if (f._dropShadow != null && f._dropShadow.knockout) {
        requiresCanvas = true;
      }
    }
    if (stroke != null || requiresCanvas) {
      surface.clear();
      if (fill != null) {
        surface.drawRect(0.0, 0.0, layoutWidth, layoutHeight, fill, stroke,
            borderRadius);
      }
    } else {
      surface.setBorderRadius(borderRadius);
      surface.setBackground(fill);
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => rectshapeElementDef;

  /** Registers component. */
  static void registerRectShape() {
    borderRadiusProperty = ElementRegistry.registerProperty("borderRadius",
        PropertyType.BORDERRADIUS, PropertyFlags.REDRAW, null, null);
    rectshapeElementDef = ElementRegistry.register("RectShape",
        Shape.shapeElementDef, [borderRadiusProperty], null);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/RectShape.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.graphics.UISurface;

import java.util.EnumSet;

/**
 * Defines a rectangle shape.
 *
 * @author ferhat@
 *
 */
public class RectShape extends Shape {

  /** Border radius Property Definition */
  public static PropertyDefinition borderRadiusPropDef = PropertySystem.register("BorderRadius",
      BorderRadius.class, RectShape.class,
      new PropertyData(BorderRadius.EMPTY, EnumSet.of(PropertyFlags.Redraw)));

  /**
   * Sets/returns border radius.
   */
  public void setBorderRadius(BorderRadius value) {
    setProperty(borderRadiusPropDef, value);
  }

  public BorderRadius getBorderRadius() {
    return (BorderRadius) getProperty(borderRadiusPropDef);
  }

  @Override
  protected void onMeasure(double availableWidth, double availableHeight) {
    double width = getWidth();
    double height = getHeight();
    setMeasuredDimension(Double.isNaN(width) ? 0 : width,
        Double.isNaN(height) ? 0 : height);
  }

  @Override
  protected void onRedraw(UISurface surface) {
    if (overridesProperty(borderRadiusPropDef)) {
      surface.drawRect(getFill(), getStroke(), new Rectangle(0, 0, getWidth(), getHeight()),
          getBorderRadius());
    } else {
      surface.drawRect(getFill(), getStroke(), new Rectangle(0, 0, getWidth(), getHeight()));
    }
  }
}
